README.md
Locale::Maketext::Gettext - Joins the gettext and Maketext frameworks
=====================================================================

Locale::Maketext::Gettext joins the [GNU gettext] and [Maketext]
frameworks.  It is a subclass of Locale::Maketext that follows the
way GNU gettext works.  It works seamlessly, both in the sense of
GNU gettext and Maketext.  As a result, you enjoy both their
advantages, and get rid of both their problems, too.

You start as a usual GNU gettext localization project:  Work on
PO files with the help of translators, reviewers and Emacs.  Turn
them into MO files with `msgfmt`.  Copy them into the appropriate
locale directory, such as
`/usr/share/locale/de/LC_MESSAGES/myapp.mo`.

Then, build your Maketext localization class, with your base class
changed from Locale::Maketext to Locale::Maketext::Gettext.  That's
all. ^_*'

[GNU gettext]: https://www.gnu.org/software/gettext/
[Maketext]: https://perldoc.perl.org/Locale::Maketext


Locale::Maketext::Gettext::Functions
------------------------------------

Locale::Maketext::Gettext::Functions is a functional
interface to Locale::Maketext::Gettext (and Locale::Maketext).
It works completely the GNU gettext way.  It plays magic to
Locale::Maketext.  No more localization class/subclasses and language
handles are required.


The `maketext` script
---------------------

The `maketext` script is a command-line interface to
Locale::Maketext::Gettext (and Locale::Maketext).  It can be used in
shell scripts, etc, to translate, maketext and return the
result.  It enables Maketext to be integrated into other programming
languages/systems, like bash/csh, python, PHP, C, etc.  It works
like the command-line program `gettext`.


System Requirements
-------------------

1. Perl, version 5.8.0 or above.  Locale::Maketext::Gettext needs the
   utf8 text internally that is only available since 5.8.0.

   You can run `perl -v` to check your current Perl version.  If you
   do not have Perl, or if you have an older version of Perl, you can
   download and install/upgrade it from the [Perl website].  For
   MS-Windows, you can download and install [Strawberry Perl] or
   [ActivePerl].

2. Required Perl modules: None.

3. Optional Perl modules: None.

[Perl website]: https://www.perl.org
[Strawberry Perl]: https://strawberryperl.com
[ActivePerl]: https://www.activestate.com/products/perl/


Download
--------

Locale::Maketext::Gettext is hosted is on…

* [Locale-Maketext-Gettext on GitHub]

* [Locale-Maketext-Gettext on MetaCPAN]

[Locale-Maketext-Gettext on GitHub]: https://github.com/imacat/Locale-Maketext-Gettext
[Locale-Maketext-Gettext on MetaCPAN]: https://metacpan.org/release/Locale-Maketext-Gettext


Install
-------

### Install with [ExtUtils::MakeMaker]

    % perl Makefile.PL
    % make
    % make test
    % make install

When running `make install`, make sure you have the privilege to write
to the installation locations.  This usually requires the `root`
privilege.

If you want to install into another location, you can set the
`PREFIX`.  For example, to install into your home when you are not
`root`:

    % perl Makefile.PL PREFIX=/home/jessica

Refer to the documentation of ExtUtils::MakeMaker for more
installation options (by running `perldoc ExtUtils::MakeMaker`).

For MS-Windows, since `make` is not universally available,
Module::Build is preferred to ExtUtils::MakeMaker.  See the
instructions below.


### Install with [Module::Build]

    % perl Build.PL
    % ./Build
    % ./Build test
    % ./Build install

When running `./Build install`, make sure you have the privilege to
write to the installation locations.  This usually requires the `root`
privilege.

If you want to install into another location, you can set the
`--prefix`.  For example, to install into your home when you are not
``root``:

    % perl Build.PL --prefix=/home/jessica

Refer to the documentation of Module::Build for more
installation options (by running `perldoc Module::Build`).


### Install with the CPAN Shell

CPAN shell takes care of ExtUtils::MakeMaker or Module::Build for you:

    % cpan Locale::Maketext::Gettext

Make sure you have the privilege to write to the installation
locations.  This usually requires the `root` privilege.  Since CPAN
shell 1.81 you can set `make_install_make_command` and
`mbuild_install_build_command` in your CPAN configuration to switch
to `root` just before install:

    % cpan
    cpan> o conf make_install_make_command "sudo make"
    cpan> o conf mbuild_install_build_command "sudo ./Build"
    cpan> install Locale::Maketext::Gettext

If you want to install into another location, you can set `makepl_arg`
and `mbuild_arg` in your CPAN configuration.  For example, to install
into your home when you are not `root`:

    % cpan
    cpan> o conf makepl_arg "PREFIX=/home/jessica"
    cpan> o conf mbuild_arg "--prefix=/home/jessica"
    cpan> install Locale::Maketext::Gettext

Refer to the documentation of cpan for more CPAN shell commands
(by running `perldoc cpan`).


### Install with the CPANPLUS Shell

CPANPLUS shell takes care of ExtUtils::MakeMaker or Module::Build for
you:

    % cpanp -i Locale::Maketext::Gettext

Make sure you have the privilege to write to the installation
locations.  This usually requires the `root` privilege.

If you want to install into another location, you can set
`makemakerflags` and `buildflags` in your CPANPLUS configuration.
For example, to install into your home when you are not `root`:

    % cpanp
    CPAN Terminal> s conf makemakerflags "PREFIX=/home/jessica"
    CPAN Terminal> s conf buildflags "--prefix=/home/jessica"
    CPAN Terminal> install Locale::Maketext::Gettext

Refer to the documentation of `cpanp` for more CPANPLUS shell
commands (by running `perldoc cpanp`).

[ExtUtils::MakeMaker]: https://metacpan.org/release/ExtUtils-MakeMaker
[Module::Build]: https://metacpan.org/release/Module-Build


Source
------

Source is now on Github.  See
https://github.com/imacat/Locale-Maketext-Gettext.


Support
-------

The Locale-Maketext-Gettext project is hosted on GitHub.  Address your
issues on the GitHub issue tracker
https://github.com/imacat/Locale-Maketext-Gettext/issues.


News, Changes and Updates
-------------------------

Refer to the Changes for changes, bug fixes, updates, new functions, etc.


Thanks
------

* Thanks to [Max Maischein] for reporting CPAN tester failures 200029,
  200332 and 200331, that helps me find the shell character escape
  problem on my test suite.

* Thanks to [Andreas Koenig] for reporting CPAN tester failures 387357
  and submitting [rt bug 23956], informing me the base class
  Locale::Maketext has updated its error handling behavior in the Perl
  5.9.

* Thanks to [Chris Travers] for suggestion to implement `pgettext()`
  in GNU gettext as `pmaketext()`.

[Max Maischein]: mailto:corion@corion.net
[Andreas Koenig]: mailto:andk@cpan.org
[rt bug 23956]: https://rt.cpan.org/Public/Bug/Display.html?id=23956
[Chris Travers]: mailto:chris.travers@gmail.com


To Do
-----

* Design a way to install the MO file through ExtUtils::MakeMaker and
  Module::Build.


Copyright
---------

    Copyright (c) 2003-2021 imacat. All rights reserved. This program is free
    software; you can redistribute it and/or modify it under the same terms
    as Perl itself.


Changes
Locale-Maketext-Gettext change log

2021-02-07	version 1.31
	 1. Updated .gitignore.
	 2. Added MANIFEST.SKIP.
	 3. MANIFEST, META.json, META.yml, SIGNATURE: Generated files
	    removed from the Git repository.
	 4. COPYING: Removed the redundant and irrelevant GPL license.
	 5. BUGS, INSTALL, README, THANKS, TODO, README.md: Replaced the
	    messy GNU-styled documents with a single Markdown README.md,
	    and updated the URLs.
	 6. Gettext.pm, Functions.pm, maketext, 11-command-line.t,
	    12-cache.t, Changes: Various fixes to avoid JetBrains IDE
	    problem detection.

2019-09-17	version 1.30
	 1. 12-cache.t: Fixed so that it works on MS-Win32.  I forgot to
	    issue binmode when copying the MO files.
	 2. 00-signature.t: Removed TEST_SIGNATURE, as this does not work
	    on everybody who does not have my GPG public key, but me.
	 3. 03-errors.t: Removed the debugging output.

2019-09-09	version 1.29
	 Fix after 10 years.
	 1. Gettext.pm: Added support to purge the lexicon cache and
	    reread the MO file on calling textdomain() when MO file is
	    updated automatically.  This enables MO file live update with
	    persistent applications like Mojolicious or mod_perl.
	 2. Source is now on Github.  See:
	    https://github.com/imacat/Locale-Maketext-Gettext

2009-06-27	version 1.28
	 Build script fix.  No code changes.
	 1. Build.PL: Removed the build_requires, as they are not
	    required by the build process, but by the test process.
	    The build process should not be blocked by them.
	    Suggested by Ryan Niebur <ryanryan52@gmail.com> and
	    Adam Kennedy <adamkennedybackup@gmail.com>
	    (and Audrey Tang).

2009-04-28	version 1.27
	 Test suite fix.  No code changes.
	 1. t/03-errors.t: Fixed the find_system_mo() subroutine to
	    exclude MO files with special characters that might be
	    considered as code by Locale::Maketext.
	 2. t/02-big-endian.t: Fixed the native-built MO file test
	    and moved the $POfile and $MOfile file name assignment
	    to the beginning, to ensure that their values are
	    assigned even if GNU gettext is not installed and
	    the test is skipped on the target test system.

2008-11-11
	 1. INSTALL: Fixed the grammar, changed "none" to "None.".

2008-04-22	version 1.26
	Documentation fix.  No code changes.
	 1. bin/maketext: Moved to script/maketext.

2008-04-21
	 1. TODO: Added.
	 2. Artistic and COPYING: Added.
	 3. INSTALL: Added.  Installation instructions moved from README
	    to INSTALL.

2008-04-11	version 1.25
	Test suite fix.  No code changes.
	 1. t/03-errors.t and t/08-f-errors.t: Added find_system_mo(), to
	    select a system MO file that we can use.

2008-02-26	version 1.24
	Test and build suite fix.  No code changes.
	 1. t/02-big-endian.t: Added creating PO at run time, in order
	    to work with GNU gettext < 0.15 whose msgfmt does not support
	    msgctxt yet.  This should work with both older and newer
	    GNU gettext.
	 2. t/test_native.po: Removed.  It is now generated by
	    t/02-big-endian.t at run time.
	 3. Makefile.PL: Added "clean" to clean-up possible existing
	    test_native.po and test_native.mo.
	 4. Build.PL: Added "add_to_cleanup" to clean-up possible existing
	    test_native.po and test_native.mo.
	 5. Build.PL: Added "build_requires".
	 6. t/08-f-errors.t: Fixed so that it finds the newest MO file
	    found on the system, in order to avoid legacy MO files.
	    (gettext 0.10 in 1995?)
	 7. t/08-f-errors.t: Fixed test 38 so that it skips in the eval()
	    block, and dumps the error on failure.
	 8. Changes: Added several missing change notes in version 1.23.

2008-02-19	version 1.23
	Added support for GNU gettext pgettext() as pmaketext(), to
	translate text messages in a particular context.
	 1. Gettext.pm: Added the pmaketext() method as an equivalent of
	    the pgettext() function in GNU gettext, to look up the text
	    message in a particular context, according to the suggestion
	    by Chris Travers.
	 2. Functions.pm: Added the pmaketext() function as an equivalent
	    of the pgettext() function in GNU gettext, and the dpmaketext()
	    function as an equivalent of the dpgettext() function in GNU
	    gettext, to look up the text message in a particular context,
	    according to the suggestion by Chris Travers.
	 3. Gettext.pm and Functions.pm: Fixed documentation and comments.
	    Changed "perl" as "Perl".  Removed contractions and replaced
	    them with full sayings, in order to work with syntax
	    highlighting.  Removed emotional marks for the same reason.
	 4. Makefile.PL: Added conditional checks for whether LICENSE is
	    available in ExtUtils::MakeMaker.
	 5. Gettext.pm and Functions.pm: Added xgettext keyword notation
	    for dmaketext(), pmaketext() and dpmaketext().
	 6. t/locale/*/LC_MESSAGES/*.mo, t/test_native.po: Added messages
	    with contexts for pmaketext() testings.
	 7. t/*.t: Added tests for pmaketext() and dpmaketext().
	 8. Gettext.pm: Fixed the failure_handler_auto() method to remove
	    the message context from the message key before sending to
	    the failure lexicon.  The failure_handler_auto() method
	    differs from its Locale::Maketext parent from now on.

2007-03-28	version 1.22
	 1. t/*-f-*.t: Moved Locale::Maketext::Gettext::Functions::_reset()
	    from the end of each set of tests, to the front of each set of
	    tests.  If at the end of the test set, it would not be run if
	    the eval{} block fails.
	 2. Gettext.pm, Functions.pm and maketext: Updated the URL of the
	    GNU gettext manual in the their POD documents.
	 3. Functions.pm: The example in its POD document is updated to show
	    more of the nature of Locale::Maketext.

2007-03-28	version 1.21
	 1. 04-encodings.t and 09-f-encodings.t: Inserted test#26 to test the
	    $@ error message.  It is an encode() error that is not affected
	    by the Locale::Maketext $@ change with "local $@".
	 2. 06-racing.t: Added $LOOKUP_FAILURE as look-up failure flag.  Added

Makefile.PL
#! /usr/bin/perl -w
require 5.008;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME        => "Locale::Maketext::Gettext",
    VERSION     => "1.31",
    ABSTRACT    => "Joins gettext and Maketext frameworks",
    AUTHOR      => "imacat <imacat\@mail.imacat.idv.tw>",
    PREREQ_PM   => { },
    LICENSE     => "perl",
    SIGN        => 1,
    
    EXE_FILES   => [ "script/maketext" ],
    dist        => {
        COMPRESS    => "gzip -9",
        SUFFIX      => ".gz",
    },
    clean		=> {
        FILES		=> "t/test_native.po t/locale/en/LC_MESSAGES/test_native.mo",
    },

    META_MERGE  => {
        "meta-spec" => { version => 2 },
        resources   => {
            homepage    => "https://metacpan.org/release/Locale-Maketext-Gettext",
            repository  => {
                type    => "git",
                url     => "git://github.com/imacat/Locale-Maketext-Gettext.git",
                web     => "https://github.com/imacat/Locale-Maketext-Gettext",
            },
            bugtracker  => {
                "web"   => "https://github.com/imacat/Locale-Maketext-Gettext/issues"
            }
        },
    },
);

__END__


Build.PL
#! /usr/bin/perl -w
require 5.008;
use strict;
use warnings FATAL => 'all';
use Module::Build;

my $build = Module::Build->new(
    dist_name       => "Locale-Maketext-Gettext",
    dist_version    => "1.31",
    dist_abstract   => "Joins gettext and Maketext frameworks",
    dist_author     => "imacat <imacat\@mail.imacat.idv.tw>",
    license         => "perl",
    sign            => 1,
    
    script_files    => [ "script/maketext" ],
    requires        => {
        "perl"          => "5.8.0",
    },
    add_to_cleanup	=> [ "t/test_native.po", "t/locale/en/LC_MESSAGES/test_native.mo" ],

    meta_merge  => {
        "meta-spec" => { version => 2 },
        resources   => {
            homepage    => "https://metacpan.org/release/Locale-Maketext-Gettext",
            repository  => {
                type    => "git",
                url     => "git://github.com/imacat/Locale-Maketext-Gettext.git",
                web     => "https://github.com/imacat/Locale-Maketext-Gettext",
            },
            bugtracker  => {
                "web"   => "https://github.com/imacat/Locale-Maketext-Gettext/issues"
            }
        },
    },
);

$build->create_build_script;

__END__
